File: src/rust/norm/src/object.rs
use crate::error::{Error, Result, check_handle};
use crate::types::*;
use crate::node::Node;
use norm_sys::*;
use std::slice;
use std::os::raw::c_char;

/// NORM object handle with RAII semantics.
///
/// A Object represents a NORM transmission object, which can be a file, data, or stream.
///
/// When dropped, the object will be released by calling `NormObjectRelease`.
#[derive(Debug)]
pub struct Object {
    /// The raw NORM object handle
    handle: NormObjectHandle,
    /// Whether this object is owned by us
    owned: bool,
}

impl Object {
    /// Create a new Object from a raw NORM object handle
    ///
    /// # Arguments
    /// * `handle` - The raw NORM object handle
    ///
    /// # Returns
    /// A new Object
    pub(crate) fn from_handle(handle: NormObjectHandle) -> Self {
        Object { handle, owned: true }
    }

    /// Create a new Object from a raw NORM object handle, without taking ownership
    ///
    /// # Arguments
    /// * `handle` - The raw NORM object handle
    ///
    /// # Returns
    /// A new Object that will not release the handle when dropped
    pub fn from_handle_unowned(handle: NormObjectHandle) -> Self {
        Object { handle, owned: false }
    }

    /// Get the type of the object
    ///
    /// # Returns
    /// The object type
    pub fn get_type(&self) -> ObjectType {
        let obj_type = unsafe { NormObjectGetType(self.handle) };
        ObjectType::from(obj_type)
    }

    /// Check if the object has info data
    ///
    /// # Returns
    /// `true` if the object has info data, `false` otherwise
    pub fn has_info(&self) -> bool {
        unsafe { NormObjectHasInfo(self.handle) }
    }

    /// Get the info data associated with the object
    ///
    /// # Returns
    /// The info data as a byte vector, or an error if the info data could not be retrieved
    pub fn get_info(&self) -> Result<Vec<u8>> {
        if !self.has_info() {
            return Ok(Vec::new());
        }

        let info_len = unsafe { NormObjectGetInfoLength(self.handle) };
        if info_len == 0 {
            return Ok(Vec::new());
        }

        let mut buffer = vec![0u8; info_len as usize];
        let bytes_read = unsafe {
            NormObjectGetInfo(
                self.handle,
                buffer.as_mut_ptr() as *mut c_char,
                info_len,
            )
        };

        if bytes_read == 0 {
            Err(Error::OperationFailed("Failed to get info data".to_string()))
        } else {
            buffer.truncate(bytes_read as usize);
            Ok(buffer)
        }
    }

    /// Get the size of the object in bytes
    ///
    /// # Returns
    /// The size of the object in bytes
    pub fn size(&self) -> Size {
        unsafe { NormObjectGetSize(self.handle) }
    }

    /// Get the number of bytes still pending transmission
    ///
    /// # Returns
    /// The number of bytes still pending transmission
    pub fn bytes_pending(&self) -> Size {
        unsafe { NormObjectGetBytesPending(self.handle) }
    }

    /// Access the data of a NORM_OBJECT_DATA object
    ///
    /// # Returns
    /// A slice containing the data, or an error if the data could not be accessed
    pub fn access_data(&self) -> Result<&[u8]> {
        if self.get_type() != ObjectType::Data {
            return Err(Error::InvalidParameter);
        }

        let data_ptr = unsafe { NormDataAccessData(self.handle) };
        if data_ptr.is_null() {
            return Err(Error::NullPointer);
        }

        let size = self.size() as usize;
        let data_slice = unsafe { slice::from_raw_parts(data_ptr as *const u8, size) };
        Ok(data_slice)
    }

    /// Cancel the object
    ///
    /// This aborts transmission or reception of the object.
    pub fn cancel(&self) {
        unsafe { NormObjectCancel(self.handle) };
    }

    /// Get the sender of the object
    ///
    /// # Returns
    /// The sender node, or an error if the sender could not be retrieved
    pub fn get_sender(&self) -> Result<Node> {
        let sender = unsafe { NormObjectGetSender(self.handle) };
        unsafe { check_handle(sender, NORM_NODE_INVALID) }
            .map(Node::from_handle_unowned)
    }

    /// Get the raw NORM object handle
    ///
    /// # Returns
    /// The raw NORM object handle
    pub(crate) fn handle(&self) -> NormObjectHandle {
        self.handle
    }

    /// Retain an object handle, incrementing its reference count
    pub fn retain(&self) {
        unsafe { NormObjectRetain(self.handle) };
    }

    /// Explicitly release an object handle, decrementing its reference count
    ///
    /// This is called automatically when the object is dropped, if it is owned.
    pub fn release(&self) {
        unsafe { NormObjectRelease(self.handle) };
    }

    // Stream-specific operations

    /// Write data to a stream object
    ///
    /// # Arguments
    /// * `data` - The data to write to the stream
    ///
    /// # Returns
    /// The number of bytes written, or an error if the operation failed
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_write(&self, data: &[u8]) -> Result<usize> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        let bytes_written = unsafe {
            NormStreamWrite(
                self.handle,
                data.as_ptr() as *const c_char,
                data.len() as u32,
            )
        };

        Ok(bytes_written as usize)
    }

    /// Flush a stream object
    ///
    /// # Arguments
    /// * `eom` - Whether to mark this as end-of-message
    /// * `flush_mode` - The flush mode to use
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_flush(&self, eom: bool, flush_mode: crate::types::FlushMode) -> Result<()> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        unsafe {
            NormStreamFlush(
                self.handle,
                eom,
                flush_mode.into(),
            );
        }

        Ok(())
    }

    /// Mark end-of-message on a stream object
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_mark_eom(&self) -> Result<()> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        unsafe { NormStreamMarkEom(self.handle) };
        Ok(())
    }

    /// Close a stream object
    ///
    /// # Arguments
    /// * `graceful` - Whether to close gracefully (wait for pending data) or immediately
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_close(&self, graceful: bool) -> Result<()> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        unsafe { NormStreamClose(self.handle, graceful) };
        Ok(())
    }

    /// Read data from a stream object
    ///
    /// # Arguments
    /// * `buffer` - The buffer to read data into
    ///
    /// # Returns
    /// A tuple of (bytes_read, data_slice) or an error if the operation failed
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    /// Returns `Error::OperationFailed` if the read operation failed
    pub fn stream_read(&self, buffer: &mut [u8]) -> Result<usize> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        let mut num_bytes = buffer.len() as u32;
        let success = unsafe {
            NormStreamRead(
                self.handle,
                buffer.as_mut_ptr() as *mut c_char,
                &mut num_bytes,
            )
        };

        if success {
            Ok(num_bytes as usize)
        } else {
            Err(Error::OperationFailed("Failed to read from stream".to_string()))
        }
    }

    /// Check if a stream has available space for writing
    ///
    /// # Returns
    /// `true` if the stream has vacancy for writing, `false` otherwise
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_has_vacancy(&self) -> Result<bool> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        let has_vacancy = unsafe { NormStreamHasVacancy(self.handle) };
        Ok(has_vacancy)
    }

    /// Seek to the start of the next message in a stream
    ///
    /// # Returns
    /// `true` if a message start was found, `false` otherwise
    ///
    /// # Errors
    /// Returns `Error::InvalidParameter` if this is not a stream object
    pub fn stream_seek_msg_start(&self) -> Result<bool> {
        if self.get_type() != ObjectType::Stream {
            return Err(Error::InvalidParameter);
        }

        let found = unsafe { NormStreamSeekMsgStart(self.handle) };
        Ok(found)
    }
}

impl Drop for Object {
    fn drop(&mut self) {
        if self.owned {
            unsafe { NormObjectRelease(self.handle) };
        }
    }
}

File: src/go/norm/object.go
package norm

import (
	"fmt"
	"unsafe"

	"github.com/USNavalResearchLaboratory/norm/src/go/normsys"
)

// Object is a NORM transport object (data, file, or stream).
//
// Ownership: objects obtained from send operations (Session.DataEnqueue,
// FileEnqueue, StreamOpen) are "owned" — Close releases the underlying NORM
// reference. Objects obtained from events (Event.Object) are "unowned" borrows
// valid only during event handling; Close is a no-op on them. To keep an unowned
// object alive past the event, call Retain (which returns an owned wrapper).
type Object struct {
	handle normsys.ObjectHandle
	owned  bool
}

// Handle exposes the raw normsys handle for use with the low-level package.
func (o *Object) Handle() normsys.ObjectHandle { return o.handle }

// Close releases the object's NORM reference if this wrapper owns it.
func (o *Object) Close() {
	if o.owned && o.handle != normsys.ObjectInvalid {
		normsys.ObjectRelease(o.handle)
		o.handle = normsys.ObjectInvalid
	}
}

// Retain increments the object's reference count and returns an owned wrapper
// whose Close will release that reference. Use it to keep a received (unowned)
// object valid beyond the event that delivered it.
func (o *Object) Retain() *Object {
	normsys.ObjectRetain(o.handle)
	return &Object{handle: o.handle, owned: true}
}

// Type returns the object's type.
func (o *Object) Type() ObjectType { return normsys.ObjectGetType(o.handle) }

// Size returns the object's total size in bytes.
func (o *Object) Size() int64 { return normsys.ObjectGetSize(o.handle) }

// BytesPending returns the number of bytes not yet sent/received.
func (o *Object) BytesPending() int64 { return normsys.ObjectGetBytesPending(o.handle) }

// HasInfo reports whether the object carries NORM_INFO content.
func (o *Object) HasInfo() bool { return normsys.ObjectHasInfo(o.handle) }

// Info returns a copy of the object's NORM_INFO content, or nil if none.
func (o *Object) Info() []byte { return normsys.ObjectGetInfo(o.handle) }

// Sender returns an (unowned) wrapper for the remote sender of a received
// object, or nil if none.
func (o *Object) Sender() *Node {
	h := normsys.ObjectGetSender(o.handle)
	if h == normsys.NodeInvalid {
		return nil
	}
	return &Node{handle: h, owned: false}
}

// Cancel aborts transmission or reception of the object.
func (o *Object) Cancel() { normsys.ObjectCancel(o.handle) }

// Data returns a copy of a received data object's payload. It is valid to call
// only on ObjectData objects; callers should typically check Type first.
func (o *Object) Data() ([]byte, error) {
	if o.Type() != ObjectData {
		return nil, fmt.Errorf("%w: Data requires ObjectData, got %s", ErrWrongObjectType, ObjectTypeString(o.Type()))
	}
	ptr := normsys.DataAccessData(o.handle)
	size := o.Size()
	if ptr == nil || size <= 0 {
		return nil, nil
	}
	// Copy out of NORM-managed memory into a Go slice; the source is freed when
	// the object is released, so we must not alias it.
	out := make([]byte, int(size))
	copy(out, unsafe.Slice((*byte)(ptr), int(size)))
	return out, nil
}

// --- stream operations (valid only on ObjectStream objects) ---

func (o *Object) requireStream(op string) error {
	if o.Type() != ObjectStream {
		return fmt.Errorf("%w: %s requires ObjectStream, got %s", ErrWrongObjectType, op, ObjectTypeString(o.Type()))
	}
	return nil
}

// StreamWrite writes to a send stream, returning bytes accepted (may be < len(p)
// under flow control).
func (o *Object) StreamWrite(p []byte) (int, error) {
	if err := o.requireStream("StreamWrite"); err != nil {
		return 0, err
	}
	return int(normsys.StreamWrite(o.handle, p)), nil
}

// StreamRead reads from a receive stream into p. inSync is false when a stream
// break was detected (see NormStreamRead semantics); the caller should then use
// StreamSeekMsgStart to resynchronize.
func (o *Object) StreamRead(p []byte) (n int, inSync bool, err error) {
	if err := o.requireStream("StreamRead"); err != nil {
		return 0, false, err
	}
	num, sync := normsys.StreamRead(o.handle, p)
	return int(num), sync, nil
}

// StreamFlush flushes buffered stream data. If eom is true it marks end-of-message.
func (o *Object) StreamFlush(eom bool, mode FlushMode) error {
	if err := o.requireStream("StreamFlush"); err != nil {
		return err
	}
	normsys.StreamFlush(o.handle, eom, mode)
	return nil
}

// StreamSetAutoFlush configures automatic flushing behavior for a send stream.
func (o *Object) StreamSetAutoFlush(mode FlushMode) error {
	if err := o.requireStream("StreamSetAutoFlush"); err != nil {
		return err
	}
	normsys.StreamSetAutoFlush(o.handle, mode)
	return nil
}

// StreamSetPushEnable controls whether writes push through the buffer aggressively.
func (o *Object) StreamSetPushEnable(enable bool) error {
	if err := o.requireStream("StreamSetPushEnable"); err != nil {
		return err
	}
	normsys.StreamSetPushEnable(o.handle, enable)
	return nil
}

// StreamHasVacancy reports whether the send stream buffer has room for more data.
func (o *Object) StreamHasVacancy() bool { return normsys.StreamHasVacancy(o.handle) }

// StreamMarkEom marks the end of the current message in a send stream.
func (o *Object) StreamMarkEom() error {
	if err := o.requireStream("StreamMarkEom"); err != nil {
		return err
	}
	normsys.StreamMarkEom(o.handle)
	return nil
}

// StreamSeekMsgStart advances a receive stream to the start of the next message
// after a break. Returns true if a message start was found.
func (o *Object) StreamSeekMsgStart() bool { return normsys.StreamSeekMsgStart(o.handle) }

// StreamClose closes a stream. If graceful is true it flushes and signals
// end-of-stream to receivers before closing.
func (o *Object) StreamClose(graceful bool) { normsys.StreamClose(o.handle, graceful) }

// --- file operations (valid only on ObjectFile objects) ---

// FileName returns the local path of a received file object.
func (o *Object) FileName() string { return normsys.FileGetName(o.handle) }

// FileRename moves/renames a received file object.
func (o *Object) FileRename(name string) error {
	return boolErr(normsys.FileRename(o.handle, name), "NormFileRename")
}
